src/main/java/com/bytecoders/pharmaid/service/SharingPermissionService.java

```diff
@@ -0,0 +1,181 @@
+package com.bytecoders.pharmaid.service;
+
+import static com.bytecoders.pharmaid.util.SecurityUtils.getCurrentUserId;
+
+import com.bytecoders.pharmaid.exception.NotAuthorizedException;
+import com.bytecoders.pharmaid.exception.ResourceNotFoundException;
+import com.bytecoders.pharmaid.exception.UserNotFoundException;
+import com.bytecoders.pharmaid.repository.SharingPermissionRepository;
+import com.bytecoders.pharmaid.repository.UserRepository;
+import com.bytecoders.pharmaid.repository.model.SharingPermission;
+import com.bytecoders.pharmaid.repository.model.SharingPermissionStatus;
+import com.bytecoders.pharmaid.repository.model.SharingRequest;
+import com.bytecoders.pharmaid.repository.model.User;
+import java.util.Date;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.access.AccessDeniedException;
+import org.springframework.stereotype.Service;
+
+/**
+ * Service class providing helper functions to manage sharing requests and permissions. Handles
+ * creation, acceptance, denial, and revocation of sharing requests between users.
+ */
+@Slf4j
+@Service
+public class SharingPermissionService {
+
+  @Autowired private SharingPermissionRepository sharingPermissionRepository;
+
+  @Autowired private UserRepository userRepository;
+
+  /**
+   * Creates a new sharing request between users.
+   *
+   * @param ownerId The ID of the user who owns the records
+   * @param request The sharing request details
+   * @return The created sharing permission
+   * @throws UserNotFoundException if either owner or requester is not found
+   * @throws IllegalArgumentException if invalid request parameters
+   */
+  public SharingPermission createSharingRequest(String ownerId, SharingRequest request) {
+    User owner =
+        userRepository
+            .findById(ownerId)
+            .orElseThrow(() -> new UserNotFoundException("User not found with ID: " + ownerId));
+
+    String requesterId = getCurrentUserId();
+
+    if (ownerId.equals(requesterId)) {
+      throw new IllegalArgumentException("Cannot create sharing permission with yourself");
+    }
+
+    User requester =
+        userRepository
+            .findById(requesterId)
+            .orElseThrow(() -> new UserNotFoundException("User not found with ID: " + requesterId));
+
+    // Check if permission already exists
+    boolean permissionExists =
```

Comment:
I would argue this is redundant for idempotency reasons. We should just not create if already exists and return the same object as if we created it

Reply:
this is to avoid spam duplicate permission requests from a user

Reply:
Yes, but it's a totally valid reason to resend a permission request. if there's one pending, just return the existing one.

Reply:
makes sense